client/client.go

```diff
@@ -0,0 +1,479 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	apimachinerymeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/ratelimiter"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	ErrNotStarted   error = errors.New("DynamicWatcher must be started to perform this action")
+	ErrInvalidInput error = errors.New("invalid input provided")
+)
+
+type Reconciler interface {
+	// Reconcile is called whenever an object has started being watched (if it exists) as well as when it has changed
+	// (added, modified, or deleted). If an error is returned, the request is requeued.
+	Reconcile(ctx context.Context, watcher ObjectIdentifier) (reconcile.Result, error)
+}
+
+// Options are the arguments for creating a new DynamicWatcher.
+type Options struct {
+	// RateLimiter is used to limit how frequently requests may be queued.
+	// Defaults to client-go's MaxOfRateLimiter (like controller-runtime does) which has both overall and per-item rate
+	// limiting. The overall is a token bucket and the per-item is exponential.
+	RateLimiter ratelimiter.RateLimiter
+}
+
+// ObjectIdentifier identifies an object from the Kubernetes API.
+type ObjectIdentifier struct {
+	Group     string
+	Version   string
+	Kind      string
+	Namespace string
+	Name      string
+}
+
+type k8sObject interface {
+	GroupVersionKind() schema.GroupVersionKind
+	GetName() string
+	GetNamespace() string
+}
+
+// ToObjectIdentifer will convert a Kubernetes object from the client-go library to an ObjectIdentifier.
+func ToObjectIdentifer(object k8sObject) ObjectIdentifier {
+	return ObjectIdentifier{
+		Group:     object.GroupVersionKind().Group,
+		Version:   object.GroupVersionKind().Version,
+		Kind:      object.GroupVersionKind().Kind,
+		Namespace: object.GetNamespace(),
+		Name:      object.GetName(),
+	}
+}
```

Comment:
I was mainly using it for the tests but had thought it might be useful to consumers but as you point out, controller-runtime's cache has a different way of representing Kubernetes objects compared to client-go, so I'll just make it a test function.